File: sdk/samples/go/search-and-filter.go
// 注: 本SDKは情報検索のみ。税理士法 §52 により、個別税務助言は税理士にご相談ください。
//
// 税務会計AI — Go paginated search + filter chain
// ----------------------------------------------------------
// Run: `go run search-and-filter.go`  (Go 1.21+; stdlib only)
// Demonstrates:
//   - paginating /v1/programs/search with limit/offset
//   - filtering by prefecture + tier + funding_purpose
//   - 429 retry-after + 5xx exponential backoff

package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"time"
)

const (
	baseURL  = "https://api.zeimu-kaikei.ai/v1"
	pageSize = 20
	maxPages = 3 // cap to spare anonymous quota (50/月)
)

type Program struct {
	UnifiedID       string  `json:"unified_id"`
	PrimaryName     string  `json:"primary_name"`
	Tier            string  `json:"tier"`
	AmountMaxManYen *float64 `json:"amount_max_man_yen"`
}

type SearchResp struct {
	Total   int       `json:"total"`
	Limit   int       `json:"limit"`
	Offset  int       `json:"offset"`
	Results []Program `json:"results"`
}

func describeStatus(code int, body string) string {
	switch {
	case code == 401:
		return "auth failed: ZEIMU_KAIKEI_API_KEY missing or invalid"
	case code == 403:
		return "forbidden: key revoked or quota exhausted"
	case code == 429:
		return "rate limited (anon = 50/月; auth = burst limit)"
	case code == 404:
		return "not found"
	case code >= 500:
		return fmt.Sprintf("server error %d: try again later", code)
	}
	return fmt.Sprintf("HTTP %d: %s", code, body)
}

func call(path string, params url.Values, attempt int) (*SearchResp, error) {
	u, _ := url.Parse(baseURL + path)
	u.RawQuery = params.Encode()

	req, _ := http.NewRequest("GET", u.String(), nil)
	req.Header.Set("Accept", "application/json")
	if key := os.Getenv("ZEIMU_KAIKEI_API_KEY"); key != "" {
		req.Header.Set("X-API-Key", key)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("transport: %w", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	if resp.StatusCode == 429 && attempt < 2 {
		retry, _ := strconv.Atoi(resp.Header.Get("Retry-After"))
		if retry == 0 {
			retry = 1
		}
		fmt.Fprintf(os.Stderr, "    retry in %ds ...\n", retry)
		time.Sleep(time.Duration(retry) * time.Second)
		return call(path, params, attempt+1)
	}
	if resp.StatusCode >= 500 && attempt < 2 {
		wait := time.Duration(500*(1<<attempt)) * time.Millisecond
		fmt.Fprintf(os.Stderr, "    server %d, backing off %v\n", resp.StatusCode, wait)
		time.Sleep(wait)
		return call(path, params, attempt+1)
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf(describeStatus(resp.StatusCode, string(body)))
	}

	var out SearchResp
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, fmt.Errorf("parse: %w", err)
	}
	return &out, nil
}

func main() {
	// Filter chain: keyword + tier S/A. Broad enough that anonymous tier
	// sees pagination kick in; tighten with prefecture / funding_purpose
	// for narrower queries.
	filters := url.Values{}
	filters.Set("q", "省エネ")
	filters.Add("tier", "S")
	filters.Add("tier", "A")

	auth := "anonymous"
	if os.Getenv("ZEIMU_KAIKEI_API_KEY") != "" {
		auth = "authenticated"
	}
	fmt.Println("Filters:", filters.Encode())
	fmt.Println("API base:", baseURL)
	fmt.Println("Auth:", auth)
	fmt.Println()

	totalSeen := 0
	pageNum := 0
	for offset := 0; pageNum < maxPages; offset += pageSize {
		p := url.Values{}
		for k, v := range filters {
			p[k] = v
		}
		p.Set("limit", strconv.Itoa(pageSize))
		p.Set("offset", strconv.Itoa(offset))

		page, err := call("/programs/search", p, 0)
		if err != nil {
			fmt.Fprintln(os.Stderr, "ERROR:", err)
			os.Exit(1)
		}
		pageNum++
		fmt.Printf("--- page %d (offset=%d, total=%d) ---\n", pageNum, offset, page.Total)
		for _, prog := range page.Results {
			amt := "金額未定"
			if prog.AmountMaxManYen != nil {
				amt = fmt.Sprintf("%.0f万円", *prog.AmountMaxManYen)
			}
			name := prog.PrimaryName
			if len(name) > 50 {
				name = name[:50]
			}
			fmt.Printf("  %s [%s] %s  %s\n", prog.UnifiedID, prog.Tier, amt, name)
		}
		totalSeen += len(page.Results)
		if len(page.Results) < pageSize || offset+pageSize >= page.Total {
			break
		}
	}

	fmt.Printf("\nFetched %d programs across %d page(s).\n", totalSeen, pageNum)
	if os.Getenv("ZEIMU_KAIKEI_API_KEY") != "" {
		fmt.Printf("Cost: %d req × ¥3 = ¥%d (税抜)\n", pageNum, pageNum*3)
	}
}

File: sdk/samples/javascript/search-and-filter.js
// 注: 本SDKは情報検索のみ。税理士法 §52 により、個別税務助言は税理士にご相談ください。
//
// 税務会計AI — Node.js paginated search + filter chain
// ----------------------------------------------------------
// Run: `node search-and-filter.js`  (Node 18+; zero deps)
// Demonstrates:
//   - paginating /v1/programs/search with limit/offset
//   - filtering by tier + prefecture + funding_purpose
//   - per-page error handling with HTTP 401/429/5xx coverage

const BASE_URL = "https://api.zeimu-kaikei.ai/v1";
const API_KEY = process.env.ZEIMU_KAIKEI_API_KEY || null;
const PAGE_SIZE = 20;
const MAX_PAGES = 3; // cap so anonymous tier doesn't burn all 50/月

function describeError(status, body) {
  if (status === 401) return "auth failed: ZEIMU_KAIKEI_API_KEY missing or invalid";
  if (status === 403) return "forbidden: key revoked or quota exhausted";
  if (status === 429) return "rate limited (anon = 50/月; auth = burst limit)";
  if (status === 404) return "not found: check unified_id or path";
  if (status >= 500) return `server error ${status}: try again later`;
  return `HTTP ${status}: ${body}`;
}

async function call(path, params = {}, attempt = 0) {
  const url = new URL(BASE_URL + path);
  for (const [k, v] of Object.entries(params)) {
    if (Array.isArray(v)) v.forEach((x) => url.searchParams.append(k, x));
    else if (v !== null && v !== undefined) url.searchParams.set(k, String(v));
  }
  const headers = { Accept: "application/json" };
  if (API_KEY) headers["X-API-Key"] = API_KEY;

  const res = await fetch(url, { headers });

  if (res.status === 429 && attempt < 2) {
    const retry = parseInt(res.headers.get("retry-after") || "1", 10);
    console.warn(`    retry in ${retry}s ...`);
    await new Promise((r) => setTimeout(r, retry * 1000));
    return call(path, params, attempt + 1);
  }
  if (res.status >= 500 && attempt < 2) {
    const wait = 0.5 * 2 ** attempt;
    console.warn(`    server ${res.status}, backing off ${wait}s`);
    await new Promise((r) => setTimeout(r, wait * 1000));
    return call(path, params, attempt + 1);
  }
  if (!res.ok) throw new Error(describeError(res.status, await res.text()));
  return res.json();
}

async function* paginate(path, baseParams) {
  let offset = 0;
  for (let page = 0; page < MAX_PAGES; page++) {
    const data = await call(path, { ...baseParams, limit: PAGE_SIZE, offset });
    yield data;
    if (data.results.length < PAGE_SIZE) return; // last page
    if (offset + PAGE_SIZE >= data.total) return;
    offset += PAGE_SIZE;
  }
}

async function main() {
  // Filter chain: keyword + tier S/A (broad enough that anonymous tier
  // sees pagination kick in; tighten with prefecture / funding_purpose
  // for narrower queries).
  const filters = {
    q: "省エネ",
    tier: ["S", "A"],
  };
  console.log("Filters:", JSON.stringify(filters));
  console.log("API base:", BASE_URL);
  console.log("Auth:", API_KEY ? "authenticated" : "anonymous");
  console.log("");

  let pageNum = 0;
  let totalSeen = 0;
  for await (const page of paginate("/programs/search", filters)) {
    pageNum++;
    console.log(`--- page ${pageNum} (offset=${(pageNum - 1) * PAGE_SIZE}, total=${page.total}) ---`);
    for (const p of page.results) {
      const amt = p.amount_max_man_yen ? `${p.amount_max_man_yen}万円` : "金額未定";
      console.log(`  ${p.unified_id} [${p.tier}] ${amt}  ${p.primary_name.slice(0, 50)}`);
    }
    totalSeen += page.results.length;
  }
  console.log(`\nFetched ${totalSeen} programs across ${pageNum} page(s).`);
  if (API_KEY) console.log(`Cost: ${pageNum} req × ¥3 = ¥${pageNum * 3} (税抜)`);
}

main().catch((e) => {
  console.error("ERROR:", e.message);
  process.exit(1);
});
